# Capstone Week 1

This notebook will be mainly used for the capstone project.

In [1]:
import pandas as pd
import numpy as np

print("Hello Capstone Project Course!")

Hello Capstone Project Course!


# Capstone Week 3

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 2. Scrape wikipedia page

Data source is: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

Approach chosen: Pandas web scraping (see also: https://pythonbasics.org/pandas-web-scraping/)

In [3]:
# make sure that beautiful is installed
!pip install lxml html5lib beautifulsoup4  

In [4]:
# load the neighborhood table from the wikipedia page into a pandas dataframe
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data_frames = pd.read_html(url)       # data_frames contains several tables
neighborhood_df = data_frames[0]      # select the first table
neighborhood_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

## 3. Create the dataframe

In [5]:
# remove all rows with Borough == "Not Assigned"
neighborhood_df.drop(neighborhood_df[neighborhood_df['Borough'] == "Not assigned"].index, inplace = True)
neighborhood_df.shape

(103, 3)

103 is also the number of neighborhoods (FSA) indicated on the wikipedia page.
So, now we have a working table.

## 4. Load geospatial data

Load longitude and latitude of each postal code and then merge with neighborhood data frame.

In [6]:
#!curl -O http://cocl.us/Geospatial_data        doesn't work with the shortened url provided
#geo_df = pd.read_csv("Geospatial_data.csv")

# try python way:
import io
import requests

url = "http://cocl.us/Geospatial_data"
geo = requests.get(url).text
geo_df = pd.read_csv(io.StringIO(geo))

geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
To_df = pd.merge(neighborhood_df, geo_df, how = 'inner', left_on = 'Postal Code', right_on = 'Postal Code')

print(To_df.shape)
To_df.head()

(103, 5)


Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## 5. Explore and cluster the neighborhoods 

#### Check number of boroughs and neighborhoods

In [8]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(To_df['Borough'].unique()),
        To_df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Create a map of Toronto with neighborhoods superimposed on top

In [9]:
# retrieve latitude and longitude of Toronto
address = 'Toronto'
geolocator = Nominatim(user_agent = "To_explorer")
location = geolocator.geocode(address)
print(location, location.latitude, location.longitude)

# create the map
map_To = folium.Map(location=[location.latitude, location.longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(To_df['Latitude'], To_df['Longitude'], To_df['Borough'], To_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_To)  
    
map_To

Toronto, Golden Horseshoe, Ontario, Canada 43.6534817 -79.3839347


#### Simplify to only those boroughs that contain the word 'Toronto'  

In [10]:
ToC_df = To_df[To_df['Borough'].str.contains('Toronto')]
ToC_df

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                         Neighbourhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            Richmond, Adelaide, King  43.650571 -79.384568  
31                        Dufferin, Dovercourt Village  43.669005 -79.442259  
36   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68     Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
69                       High Park, The Junction South  43.661608 -79.464763  
73                   North Toronto West, Lawrence Park  43.715383 -79.405678  
74                 The Annex, North Midtown, Yorkville  43.672710 -79.405678  
75                              Parkdale, Roncesvalles  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      University of Toronto, Harbord  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Moore Park, Summerhill East  43.689574 -79.383160  
84           Kensington Market, Chinatown, Grange Park  43.653206 -79.4000

In [11]:
ToC_df.shape

(39, 5)

Show map with this selection and zoom in a little

In [12]:
# create the map around the coordinates of Central Bay Street, Downtown Toronto
map_ToC = folium.Map([ToC_df.loc[24, 'Latitude'], ToC_df.loc[24, 'Longitude']], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(ToC_df['Latitude'], ToC_df['Longitude'], ToC_df['Borough'], ToC_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_ToC) 

map_ToC

(here follows a hidden cell with FourSquare credentials)

In [13]:
# @hidden_cell
CLIENT_ID = 'ZZXEETEYC15BSOGFJVZBH00C3FPJJUCA55EPSZDFHUGIZTHA' # your Foursquare ID
CLIENT_SECRET = 'BRKOSHYRXYN4FXMIBWAWZNSGTFBR04RSFG1YPCDRPHRI2PBW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

#### Get venues within a radius of 500m around the coordinates of each of the neighborhoods

In [14]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [15]:
# function to explore all neighborhoods in a data frame
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [16]:
ToC_venues = getNearbyVenues(names = ToC_df['Neighbourhood'],
                                   latitudes = ToC_df['Latitude'],
                                   longitudes = ToC_df['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [17]:
ToC_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

In [18]:
ToC_venues.shape

(1639, 7)

In [19]:
print('There are {} uniques venue categories.'.format(len(ToC_venues['Venue Category'].unique())))

There are 240 uniques venue categories.


#### Number of venues for each neighborhood

In [20]:
ToC_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton, Parkdale Village, Exhibition Place                           25   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           13   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          26   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               46   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                5   
Roselawn                                                                3   
Runnymede, Swansea                                                     35   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            50   
Stn A PO Boxes                                                         96   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth West, Riverdale                                           44   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton, Parkdale Village, Exhibition Place                            25   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

#### Analyze each neighborhood

In [21]:
# one hot encoding
ToC_onehot = pd.get_dummies(ToC_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ToC_onehot['ZZZNeighbourhood'] = ToC_venues['Neighborhood']   # 'ZZZ' as workaround to get Neighboorhood in the last column 
# use Neighbourhood spelling to avoid conflict with one_hot column Neighborhood

# move neighborhood column to the first column
fixed_columns = [ToC_onehot.columns[-1]] + list(ToC_onehot.columns[:-1])
ToC_onehot = ToC_onehot[fixed_columns]

# restore name of Neighborhood column
ToC_onehot.rename(columns = {"ZZZNeighbourhood": "Neighbourhood"}, inplace = True)

ToC_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       1     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  \
0              0                     0               0        0   
1              0                     0               0        0   
2              0                     0               0        0   
3              0                     0               0        0   
4              0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

In [22]:
ToC_onehot.shape

(1639, 241)

Simplify onehot table by grouping per neighborhood and by taking the mean of the frequency of occurance of each category

In [23]:
ToC_grouped = ToC_onehot.groupby("Neighbourhood").mean().reset_index()
ToC_grouped.head()

Neighbourhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0.000000            0.000000      0.000000        0.000000   
1  0.000000            0.000000      0.000000        0.000000   
2  0.000000            0.000000      0.000000        0.000000   
3  0.058824            0.058824      0.058824        0.117647   
4  0.000000            0.000000      0.000000        0.000000   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0         0.000000          0.000000                  0.0           0.0   
1         0.000000          0.000000                  0.0           0.0   
2         0.000000          0.000000                  0.0           0.0   
3         0.117647          0.058824                  0.0           0.0   
4         0.000000          0.000000                  0.0           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0     0.018182    0.000000                  0.0               0.0   
1       0.0     0.000000    0.000000                  0.0               0.0   
2       0.0     0.000000    0.000000                  0.0               0.0   
3       0.0     0.000000    0.000000                  0.0               0.0   
4       0.0     0.000000    0.014706                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0         0.0000        0.0         0.0    0.018182   
1                 0.0         0.0000        0.0         0.0    0.000000   
2                 0.0         0.0625        0.0         0.0    0.000000   
3                 0.0         0.0000        0.0         0.0    0.000000   
4                 0.0         0.0000        0.0         0.0    0.000000   

     Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium     Beach  \
0  0.036364   0.0  0.000000               0.0            0.018182  0.018182   
1  0.040000   0.0  0.040000               0.0            0.000000  0.000000   
2  0.000000   0.0  0.000000               0.0            0.000000  0.000000   
3  0.000000   0.0  0.058824               0.0            0.000000  0.000000   
4  0.000000   0.0  0.000000               0.0            0.000000  0.000000   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0              0.0  0.036364         0.0                 0.0  0.018182   
1              0.0  0.000000         0.0                 0.0  0.000000   
2              0.0  0.000000         0.0                 0.0  0.000000   
3              0.0  0.000000         0.0                 0.0  0.000000   
4              0.0  0.000000         0.0                 0.0  0.000000   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Brazilian Restaurant  \
0       0.000000   0.000000  0.000000            0.0                   0.0   
1       0.000000   0.040000  0.000000            0.0                   0.0   
2       0.000000   0.000000  0.000000            0.0                   0.0   
3       0.058824   0.000000  0.058824            0.0                   0.0   
4       0.000000   0.014706  0.000000            0.0                   0.0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0        0.018182   0.0000         0.000000       0.0      0.000000   
1        0.080000   0.0000         0.000000       0.0      0.000000   
2        0.000000   0.0625         0.000000       0.0      0.000000   
3        0.000000   0.0000         0.000000       0.0      0.000000   
4        0.000000   0.0000         0.029412       0.0      0.029412   

   B

In [24]:
ToC_grouped.shape

(39, 241)

Identify 10 most common venue categories per neighbourhood

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = ToC_grouped['Neighbourhood']

for ind in np.arange(ToC_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ToC_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar        Farmers Market   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service              Boutique                   Bar   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar            Restaurant           Cheese Shop   
1                   Gym            Playground               Stadium   
2               Butcher        Farmers Market  Fast Food Restaurant   
3         Boat or Ferry      Sculpture Garden         Historic Site   
4      Department Store   Japanese Restaurant           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant          Concert Hall     Basketball Stadium  
1                   Bar    Italian Restaurant                 Bakery  
2            Restaurant      Recording Studio          Auto Workshop  
3   Rental Car Location       Harbor / Marina            Coffee Shop  
4       Bubble Tea Shop          Burger Joint        Thai Restaurant

#### Cluster neighbourhoods by k-means

In [27]:
# set number of clusters
kclusters = 5

ToC_grouped_clustering = ToC_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(ToC_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [28]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

ToC_merged = ToC_df

# merge ToC_grouped with ToC_data to add latitude/longitude for each neighborhood
ToC_merged = ToC_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

ToC_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
15         M5C  Downtown Toronto                               St. James Town   
19         M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               0           Coffee Shop   
4   43.662301 -79.389494               0           Coffee Shop   
9   43.657162 -79.378937               0           Coffee Shop   
15  43.651494 -79.375418               0           Coffee Shop   
19  43.676357 -79.293031               2                   Pub   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                Bakery                   Pub   
4            Yoga Studio                 Diner            Restaurant   
9         Clothing Store                  Café       Bubble Tea Shop   
15                  Café          Cocktail Bar            Restaurant   
19     Health Food Store                 Trail          Neighborhood   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2          Breakfast Spot                  Café               Theater   
4   Portuguese Restaurant                  Park           Music Venue   
9     Japanese Restaurant        Cosmetics Shop                 Hotel   
15               Beer Bar             Gastropub   American Restaurant   
19            Yoga Studio              Dive Bar    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
2             Shoe Store               Brewery                 Restaurant  
4     Mexican Restaurant    Italian Restaurant                 Hobby Shop  
9              Bookstore           Pizza Place  Middle Eastern Restaurant  
15        Farmers Market                 Hotel        Japanese Restaurant  
19                 Diner        Discount Store        Distribution Center

#### Visualize cluster labels in a map

In [29]:
# create map
map_clusters = folium.Map(location = [ToC_df.loc[24, 'Latitude'], ToC_df.loc[24, 'Longitude']], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(ToC_merged['Latitude'], ToC_merged['Longitude'], ToC_merged['Neighbourhood'], ToC_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Show neighbourhoods per cluster

Cluster 1

In [30]:
ToC_merged.loc[ToC_merged['Cluster Labels'] == 0, ToC_merged.columns[[1] + list(range(5, ToC_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
2    Downtown Toronto               0                    Coffee Shop   
4    Downtown Toronto               0                    Coffee Shop   
9    Downtown Toronto               0                    Coffee Shop   
15   Downtown Toronto               0                    Coffee Shop   
20   Downtown Toronto               0                    Coffee Shop   
24   Downtown Toronto               0                    Coffee Shop   
25   Downtown Toronto               0                  Grocery Store   
30   Downtown Toronto               0                    Coffee Shop   
31       West Toronto               0                       Pharmacy   
36   Downtown Toronto               0                    Coffee Shop   
37       West Toronto               0                            Bar   
41       East Toronto               0               Greek Restaurant   
42   Downtown Toronto               0                    Coffee Shop   
43       West Toronto               0                           Café   
47       East Toronto               0                           Park   
48   Downtown Toronto               0                    Coffee Shop   
54       East Toronto               0                    Coffee Shop   
67    Central Toronto               0               Department Store   
68    Central Toronto               0               Sushi Restaurant   
69       West Toronto               0             Mexican Restaurant   
73    Central Toronto               0                    Coffee Shop   
74    Central Toronto               0                 Sandwich Place   
75       West Toronto               0                 Breakfast Spot   
79    Central Toronto               0                   Dessert Shop   
80   Downtown Toronto               0                           Café   
81       West Toronto               0                           Café   
84   Downtown Toronto               0  Vegetarian / Vegan Restaurant   
86    Central Toronto               0                    Coffee Shop   
87   Downtown Toronto               0                 Airport Lounge   
92   Downtown Toronto               0                    Coffee Shop   
96   Downtown Toronto               0                    Coffee Shop   
97   Downtown Toronto               0                    Coffee Shop   
99   Downtown Toronto               0                    Coffee Shop   
100      East Toronto               0             Light Rail Station   

    2nd Most Common Venue      3rd Most Common Venue  4th Most Common Venue  \
2                    Park                     Bakery                    Pub   
4             Yoga Studio                      Diner             Restaurant   
9          Clothing Store                       Café        Bubble Tea Shop   
15                   Café               Cocktail Bar             Restaurant   
20                 Bakery               Cocktail Bar         Farmers Market   
24                   Café             Sandwich Place     Italian Restaurant   
25                   Café                       Park             Restaurant   
30                   Café                        Gym             Restaurant   
31                 Bakery  Middle Eastern Restaurant                   Bank   
36               Aquarium                       Café                  Hotel   
37            Coffee Shop                Men's Store  Vietnamese Restaurant   
41            Coffee Shop         Italian Restaurant              Bookstore   
42                  Hotel                 Restaurant                   Café   
43         Breakfast Spot                  Nightclub            Coffee Shop   
47                    Gym                        Pub                Brewery   
48             Restaurant                      Hotel                   Café   
54                 Bakery                       Café              Gastropub   
67         Sandwich Place               Dance Studio                  H

Cluster 2

In [31]:
ToC_merged.loc[ToC_merged['Cluster Labels'] == 1, ToC_merged.columns[[1] + list(range(5, ToC_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61   Central Toronto               1                  Park   
91  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
61              Bus Line           Swim School    Dim Sum Restaurant   
91                 Trail          Tennis Court            Playground   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
61           Escape Room     Electronics Store  Eastern European Restaurant   
91            Donut Shop      Doner Restaurant                      Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
61   Dumpling Restaurant            Donut Shop       Doner Restaurant  
91   Dumpling Restaurant         Deli / Bodega               Dive Bar

Cluster 3

In [32]:
ToC_merged.loc[ToC_merged['Cluster Labels'] == 2, ToC_merged.columns[[1] + list(range(5, ToC_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19  East Toronto               2                   Pub     Health Food Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19                 Trail          Neighborhood           Yoga Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19              Dive Bar    Dim Sum Restaurant                 Diner   

   9th Most Common Venue 10th Most Common Venue  
19        Discount Store    Distribution Center

Cluster 4

In [33]:
ToC_merged.loc[ToC_merged['Cluster Labels'] == 3, ToC_merged.columns[[1] + list(range(5, ToC_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               3              Dive Bar   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62                Garden           Music Venue           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
62      Department Store     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62   Dumpling Restaurant            Donut Shop       Doner Restaurant

Cluster 5

In [34]:
ToC_merged.loc[ToC_merged['Cluster Labels'] == 4, ToC_merged.columns[[1] + list(range(5, ToC_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto               4            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83                 Trail           Yoga Studio          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
83           Escape Room     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
83   Dumpling Restaurant            Donut Shop       Doner Restaurant